### DISCUSSION TITLE
NextJS API HTML streaming

### DISCUSSION THREAD

**Discussion by rdylina** [vercel/next.js#46058] [Help]
Does anyone have a YouTube video or example repo of doing a POST to a nextjs API endpoint and then receiving streaming JSON back? I'm trying to wrap my head around this use case.

**Comment by l4b4r4b4b4** [comment]
Trying the same... 
I got a funny behaviour implementing a feature with streams (GPT-streaming). So I implemented it in three different ways on the backend. Where only the first works as expected...
1. with the example provided by vercel, that uses edge runtime and the following funtion struction 
```
export default handler = (req. Request): Promise<Response> => {
...
const stream = callOpenAiApi(payload)
return new Response(stream)
}
```
2. standard api route  NextApiResponse.
```
export default handler = (req. NextApiRequest, res: NextApiResponse) => {
...
const stream = callOpenAiApi(payload)
res.setHeader('Content-Type', 'application/json');
res.setHeader('Transfer-Encoding', 'chunked');

    const reader = stream.getReader();

    const processStream = async () => {
      try {
        while (true) {
          const { value, done } = await reader.read();
          if (done) {
            break;
          }
          res.write(value);
        }
      } catch (error) {
        console.error('Error reading stream:', error);
        res.status(500);
      } finally {
        res.end();
      }
    };

    await processStream();
}
```
3. with new route handlers as follows:
```
export async function POST(request: Request): Promise<Response> {
...
const stream = callOpenAiApi(payload)
return new Response(stream); OR return new NextResponse(stream);
```

In all cases the OpenAI stream gets handled in chunks as expected. Only that in the two latter it is not recognized properly in the frontend. Right now I have reverted to one implementation for the client-side part, but I have tried several different ones, with useEffect for example... Would be very grateful for clarification to first understand and after resolve the issue!

**Reply by l4b4r4b4b4** [reply]
OK. I got it working with onoe of the canary release :)
`"next": "^13.2.5-canary.26",`
